model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/JpaChangesPerformer.java

```diff
@@ -395,132 +397,151 @@ private void processUserSessionUpdate(KeycloakSession innerSession, Map.Entry<K,
+                    @Override
+                    public String getNote(String name) {
+                        return entity.getNotes().get(name);
+                    }
 
-                @Override
-                public void setNote(String name, String value) {
-                    throw new IllegalStateException("not implemented");
-                }
+                    @Override
+                    public void setNote(String name, String value) {
+                        throw new IllegalStateException("not implemented");
+                    }
 
-                @Override
-                public void removeNote(String name) {
-                    throw new IllegalStateException("not implemented");
-                }
+                    @Override
+                    public void removeNote(String name) {
+                        throw new IllegalStateException("not implemented");
+                    }
 
-                @Override
-                public Map<String, String> getNotes() {
-                    return entity.getNotes();
-                }
+                    @Override
+                    public Map<String, String> getNotes() {
+                        return entity.getNotes();
+                    }
 
-                @Override
-                public State getState() {
-                    return entity.getState();
-                }
+                    @Override
+                    public State getState() {
+                        return entity.getState();
+                    }
+
+                    @Override
+                    public void setState(State state) {
+                        throw new IllegalStateException("not implemented");
+                    }
+
+                    @Override
+                    public void restartSession(RealmModel realm, UserModel user, String loginUsername, String ipAddress, String authMethod, boolean rememberMe, String brokerSessionId, String brokerUserId) {
+                        throw new IllegalStateException("not implemented");
+                    }
+                }, entity.isOffline());
+            }
 
-                @Override
-                public void setState(State state) {
-                    throw new IllegalStateException("not implemented");
-                }
 
-                @Override
-                public void restartSession(RealmModel realm, UserModel user, String loginUsername, String ipAddress, String authMethod, boolean rememberMe, String brokerSessionId, String brokerUserId) {
-                    throw new IllegalStateException("not implemented");
-                }
-            }, entity.isOffline());
-        } else {
             PersistentUserSessionAdapter userSessionModel = (PersistentUserSessionAdapter) userSessionPersister.loadUserSession(realm, entry.getKey().toString(), entity.isOffline());
             if (userSessionModel != null) {
+                if (merged.getOperation() == SessionUpdateTask.CacheOperation.ADD) {
```

Comment:
I don't think we use `CacheOperation.ADD` anywhere in the code. Can we eliminate it to simplify things a bit?

Reply:
A `switch` may be simpler to read
```java
        switch (merged.getOperation()) {
            case REMOVE:
                remove();
                break;
            case ADD, ADD_IF_ABSENT:
                entity = load();
                if (entity == null) {
                    create();
                } else {
                    merge();
                }
                break;
            case REPLACE:
                entity = load();
                if (entity != null) {
                    merge();
                }
                break;
        }
```

Reply:
I addressed this in a refactoring. Please have a look.